tempesta_fw/sock.c

```diff
@@ -549,13 +546,16 @@ ss_tcp_process_data(struct sock *sk)
 			goto out;
 		}
 
-		__skb_unlink(skb, &sk->sk_receive_queue);
-		skb_orphan(skb);
-
 		/* Shared SKBs shouldn't be seen here. */
 		if (skb_shared(skb))
 			BUG();
 
+		__skb_unlink(skb, &sk->sk_receive_queue);
+		skb_orphan(skb);
+
+		/* @skb should be rerouted on forwarding. */
+		skb_dst_drop(skb);
```

Comment:
This is overkill as well. For example, it we consider web-server workload when no ingress skb is forwarded at all. Other example is DDoS attack when most of ingress packets are dropped as well.

I think introducing flag like `TFW_FORWARDED` to skb or passing it through function calls will do good job... However, if skb is generated then `skb_dst_drop()` is just one conditional check and its overhead is the same as for the flag checking.